### DISCUSSION TITLE
기획자, 디자이너, 개발자가 개발 합을 맞추기 위해 떠난 5일 간의 여정

### DISCUSSION THREAD

**Discussion by gorgeouseowoo** [DeveloperAcademy-POSTECH/.github#94] [Idea]
### 안녕하세요! 3기 주니어 러너 티나입니다! 🦖

PM을 꿈꾸는 제가 테크포럼에 글을 쓸 줄은 상상도 못했어요!
저는 올해 3월부터 개발 공부를 시작했습니다. 아직 개발 새싹이에요.
🔥**UIKit으로 5일만에 뷰 그리기**🔥 그 어려운걸 제가 해냅니다. ~~(저만 어려울 수도 있어요)~~

SwiftUI로 간단한 뷰만 그릴 수 있었던 제가
어떻게 5일만에 UIKit으로 테이블 뷰를 구현할 수 있었을까요?

바로 메크로 팀 **악동쿠키들과 함께 한 UIKit 스터디** 때문입니다! 
본격 개발에 들어가기 전, 기획자, 디자이너, 개발자 팀원이 개발 합을 맞추기 위해 5일 간 함께 UIKit 스터디를 진행했는데요,
이 과정에서 제가 얻은 배움과 성장을 공유하고 싶어 테크포럼에 글을 쓰게 되었습니다. 

지금부터 시작할게요!

## 이런 분에게 추천하는 글이에요

1. 처음 UIKit을 공부하는 사람
2. 개발 협업이 궁금한 사람
4. 티나를 응원하는 모든 러너

## 🟢 UIKit 스터디? 차라리 빨리 개발하는게 낫지 않아?

라고 생각하셨나요? 아닙니다!

운동을 하기 전에 부상을 예방하기 위해 스트레칭을 하듯이, 
**개발을 하기 전에도 협업을 위한 준비 운동이 필요합니다.** 

무엇보다 우리 팀은 개발에 익숙하지 않은 팀원들이 UIKit을 이해하고, 
PR문화에 익숙해지는 것이 또 하나의 목표였습니다.

그래서 우리는 작지만 확실한 스터디 목표를 세웠습니다.

1️⃣ 원하는 화면을 UIKit으로 생성
UI 제작 도구로서 UIKit의 동작 원리, 사용 방식을 학습합니다

> 최종적으로 UIKit 기반 프로젝트를 진행할 수 있습니다

2️⃣ UIKit 프로젝트에 SwiftUI를 적용
필요 시 UIKit 프로젝트에 SwiftUI를 사용합니다

> 최종적으로 개발 시 특정 화면에 유리한 레이아웃 방식을 선택할 수 있습니다

3️⃣ UI Components 별 접근성 검증
VoiceOver로 화면이 읽히는 방식을 점검합니다
그래프 속 정보를 어떤식으로 전달 해야할 지 고민합니다

> 최종적으로 눈을 감고도 사용 가능한 프로젝트를 제작 합니다

4️⃣ UnitTest / UITest의 필요를 체감하고 적절한 테스트 케이스를 작성
엣지 케이스에 대해 고민하고 공유하는 개발 문화를 구축합니다
기능 명세서를 토대로 적절한 TestCase를 뽑아낼 수 있습니다
리팩토링 시 수행되어야 하는 검증, 테스트를 자동화 합니다
개발 고민이 담긴 로그(테스트 케이스)를 통해 타인의 코드를 이해합니다

> 최종적으로 팀원의 부재에도 매끄럽게 대처할 수 있는 개발 환경을 구축합니다

5️⃣ PR을 통한 소통에 적응, 효용 이해
첫 커밋을 시작으로 PR을 생성하고 소통합니다
의견을 묻고 답하는 상황에 적응 합니다

> 최종적으로 상황 공유, 논의에 드는 시간을 단축하고 더 높은 퀄리티의 코드를 유지합니다

## 🟢 5일동안 매일매일 미션 과제를 수행하라!

보노가 과제를 내주면 이오, 로셸, 티나가 매일 과제를 수행하는 방식으로 진행했습니다. 
<img width="700" alt="image (2)" src="https://github.com/user-attachments/assets/21758414-cb85-4c92-b7fa-f4255eab2fda">

우리 앱과 비슷한 화면을 스토리보드와 코드로 구현해 보면서 단계적으로 UIKit 기본 개념을 쌓았고,
다이나믹폰트, VoiceOver 등 접근성 기능을 적용해 보면서 사용자 경험을 위한 개발도 학습했습니다. 
 

진행 방식에 대한 자세한 설명을 위해 예시를 들고 왔어요!
<img width="500" alt="image (2)" src="https://github.com/user-attachments/assets/669d3f6e-e243-40a8-8abd-20596de94807">
UIKit 스토리보드로 만든 화면을 지우고 코드 베이스로 똑같은 화면을 구현하는 3일차 과제입니다. 
(2일동안 열심히 구현한 화면을 삭제하려니 마음이 아팠어요💦😥)

### PR문화에 익숙해 지기

과제를 수행하면 ‘[Mission/#4-Tina] 3일차 미션’이라는 제목과 함께 PR을 올립니다. 
PR의 내용과 형식에 대해 특별한 규칙을 정하지 않았습니다. 
팀 모두가 PR을 생성하고, 코멘트를 남기는 것에 익숙해 지는 것이 중요했기 때문이죠.
실시간 작업 내용을 공유해도 좋고, 모르는 것에 대해 질문을 하기도 했어요.

이 과정에서 **‘좋은 PR이란, 정확하고 빠르게 읽혀야 한다**’는 것을 깨달았습니다. 
**PR은 소통을 위한 글이기 때문에 협업하는 사람이 보기 쉽게 작성해야 하는 것이죠!**

로셸과 티나의 PR을 한 번 살펴볼까요? 

<img width="500" alt="image (2)" src="https://github.com/user-attachments/assets/98f2dfa6-7791-47b4-bb45-e1a01a09c136">

<img width="500" alt="image (2)" src="https://github.com/user-attachments/assets/30bc93c4-8893-449f-adcd-efae5b0906e6">

둘다 좋은 PR이라고 생각해요! 

로셸의 PR에는 화면 구현을 위해 고민한 과정을 볼 수 있고, 
저의 PR에서는 진행한 task를 한 눈에 파악할 수 있습니다.

하지만 PR의 개수가 많아지고 내용이 길어지게 된다면, 핵심 내용을 파악하는데 시간이 오래 걸리게 되겠죠?
**이 경험을 통해 PR 작성에는 약속이 필요하고, PR 템플릿의 필요성을 느꼈어요😊**

### 물어보고, 대답하고, 성장하고!

왜 이런 과정이 중요할까요? 

- 질문을 통해 문제를 해결하고 새로운 것을 배우면서 자신감이 높아져요
- 내 질문이 다른 사람에게도 도움이 돼요
- 설명하는 사람도 개념을 한 번 더 복습할 수 있어요

개인적으로는 과제 중에서 오토레이아웃 설정이 가장 어려웠습니다. 

1시간 동안 오토레이아웃과 싸우다가
결국 코멘트를 통해 질문을 남겼습니다. 
<img width="500" alt="image (2)" src="https://github.com/user-attachments/assets/4b0282f9-b880-4734-b173-78e7609137e2">

다급한 도움요청 PR을 본 보노가
어떤 부분에서 문제가 발생했는지를 정확하게 짚어주었습니다.
<img width="500" alt="image (2)" src="https://github.com/user-attachments/assets/03c90041-e055-413a-868e-33e3c2b4582a">

보노 덕분에 문제를 해결하고, 
오토레이아웃 개념에 대해 한 번 더 공부했습니다.
<img width="500" alt="image (5)" src="https://github.com/user-attachments/assets/59a42d78-be26-4471-aead-529a643eb738">

우리는 혼자가 아닌 함께 성장하는 개발자입니다. 
**열린 소통을 하는 팀에서 많은 것을 배우고 성장할 수 있다고 생각합니다.**

**‘물어보고, 대답하고, 성장하는’ 과정을 두려워 하지 마세요!**

## 🟢 똑같은 화면을 함께 개발해야 하는 이유

똑같은 화면을 함께 개발하면서 많은 것을 배웠습니다.

- 다른 사람의 코드에서 힌트를 얻을 수 있다
- 각자의 접근 방식을 비교하며 좋은 코드를 찾아간다

### 각자의 접근 방식을 비교하며 좋은 코드를 찾아간다

함수와 클로저 방식 중 어떤 것을 선호하는지 팀원들에게 물어보았습니다.
<img width="500" alt="Screenshot 2024-11-17 at 2 37 44 AM 1@2x" src="https://github.com/user-attachments/assets/920e8112-82bc-4d9a-87cc-6d157ad04d74">

선호하는 방식과 이유에 대해 다양한 이야기를 들을 수 있었습니다.
<img width="500" alt="image (4)" src="https://github.com/user-attachments/assets/6fd7a422-4ac7-4907-85c0-31435f41ad91">

이오가 더 간결하게 적을 수 있는 방법을 알려주었습니다. 
<img width="500" alt="image (4)" src="https://github.com/user-attachments/assets/e8f4d985-c8a1-42f4-bb6a-8e52b151acc3">

이렇게 여러 접근 방식을 비교하다보면 
**자연스럽게 더 효율적이고 가독성 높은 코드**를 찾게 되는 것 같습니다. 

## 🟢 UIKit 스터디가 성공했던 이유는요,

**서로의 성장에 진심인 팀원들**이 있었기 때문입니다.

- 밤 늦게까지 코드를 고민하는 로셸 🍩
- 모든 PR에 코멘트를 달아주는 보노 🦦
- 새로운 접근 방식을 제안하는 이오 🐧
- 다양한 시도를 공유하는 티나 🦖

**형식적인 리뷰가 아닌, 진심으로 서로의 성장을 응원하는 시간**
그 과정에서 우리는 단순히 개발을 배우는 것을 넘어 서로를 이해하고 존중하는 법을 배웠습니다. 

이런 문화야말로 진정한 협업의 시작이 아닐까요?

**Comment by Rochelle0922** [comment]
악동쿠키들과 함께한 5일이 스쳐 지나가네요! 티나와 밤새 코멘트로 대화하면서 의지가 많이 되었습니다!!!! 앞으로 남은 기간도 화이팅🦖💚

**Comment by thinkySide** [comment]
많이 사용되는 기술, 컨벤션 등을 바로 적용하는 것이 아닌
그 이유를 직접 경험하며 함께 탄탄히 쌓아나가시는 협업 과정이 너무 멋있는 것 같습니다!! 🚀
